### DISCUSSION TITLE
Cryptographic details - missing clean documentation page?

### DISCUSSION THREAD

**Discussion by ArKam** [dani-garcia/vaultwarden#3723] [Q&A]
Hi team, as already told, I'm currently creating a bitwarden/vaultwarden compatible ansible plugin, however I'm facing a weird issue where I can't have a password hash to match the stored password hash.

Let me explain.

I do extracted from sqlite DB the server side password salt and password hash using:

`select hex(salt),hex(password_hash) from users where email='user_email';`

The issue here is I end up with what looks like two 64 bytes hashes, one for the salt, one for the server side password hash where according to the official bitwarden white paper PDF I'm supposed to get a salt and a 32 byte server side password hash.

I've look at how those two payloads are created on vaultwarden and if I'm not fooling myself it all come from:
https://github.com/dani-garcia/vaultwarden/blob/main/src/crypto.rs
&
https://github.com/dani-garcia/vaultwarden/blob/main/src/db/models/user.rs#L107
for the salt part.

What I can't get is, how the hash_password function can leverage pbkdf2 with hmac-sha256 with a random 64 bytes for salt and result in a 64 bytes long password hash in DB?

Did I missed something?
Is the data stored on the salt and password_hash as BLOB on DB really hex data type or something else?

As a result, I can't generate any matching password hash.
Additionally you can found a script that I use to validate the hashing/key derivation process in here:
https://gist.github.com/ArKam/e385c251dcbcdb7c1b3c99bf4c2ce187

I'm using the following parameters to test:

> email = g@g.dev
> password = secret
> salt = "D5AF4483678B7806E4B310AB76DBE47970E597F582F9F6B74E4CCA8462F058C4"
> expected_password_hash = "A1B5A53993D2CD253662375201B5DF61B14D7B630445ED328E7B245CF4C4C6FF73A5DB5DF37B2026717F9106F171D8FB588B4721710803F256144F95001B05A0"

With salt being the DB extracted salt as well as expected_password_hash value.

Thanks a lot!

**Comment by ArKam** [comment]
Anyone?

**Reply by BlackDex** [reply]
I suggest to read https://vaultwarden.discourse.group/t/please-help-generating-master-password-hash-in-server-db-to-retrieve-wifes-lost-password/2724

**Reply by ArKam** [reply]
Hi! Thank you very much for this page it indeed help a lot to confirm that the server isn’t hashing the master password but a b64 encoded master password.

Just for anyone reading this, the security white paper from 2021 IS NO LONGER relevant as the crypto did slightly changed and is actually kinda open sourced on https://www.bitwarden.com/crypto.html and their github repo.

although their typescript repo is a mess with intricated call for function from too many different places depending on which context you’re on.

PS: Vaulwarden rust sources made it way much smoother to understand a bit better what was going on plus this contributor response so cudos to you kind responder and the vaultwarden team.
